<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="link_2">

<!-- GEOMETRY -->
    <xacro:macro name="link_2_geometry_1">
        <origin rpy="0 0 0" xyz="${link_2_length/2} 0 ${link_2_height/2}" />
        <geometry>
            <box size="${link_2_length} ${link_2_width} ${link_2_height}"/>
        </geometry>
    </xacro:macro>

    <xacro:macro name="link_2_geometry_2">
        <origin rpy="1.57075 0 0" xyz="${link_2_length} 0 ${link_2_height/2}" />
        <geometry>
            <cylinder length="${link_2_width}" radius="${link_2_height/2}"/>
        </geometry>
    </xacro:macro>

    <!-- VISUAL -->
    <xacro:macro name="link_2_visual">
        <origin xyz="0.34 0.093 -0.523" rpy="1.57075 0 -1.45"/>
        <geometry>
            <mesh filename="package://phobos_meshes/rover_stl/stopien_4.STL" scale="0.001 0.001 0.001"/>
        </geometry>
    </xacro:macro>


<!-- MATERIAL -->
    <xacro:macro name="link_2_material">
        <material name="link_2_color">
            <color rgba=".4 .4 .8 1"/>
        </material>
    </xacro:macro>

<!-- LINK -->
    <xacro:macro name="link_2" params="name parent x y z">

        <link name="${name}">
            <visual>
                <xacro:link_2_visual/>
                <xacro:link_2_material/>
            </visual>

            <collision>
                <xacro:link_2_geometry_1/>
            </collision>
            <collision>
                <xacro:link_2_geometry_2/>
            </collision>
            <inertial>
                <origin rpy="0 0 0" xyz="${link_2_length/2} 0 ${link_2_height/2}" />
                <mass value="1"/>
                <inertia ixx=".003" ixy="0.000" ixz="0.000" iyy=".003" iyz="0.000" izz=".003"/>
            </inertial>
        </link>
        <gazebo reference="${name}">
            <mu1> .2 </mu1>
            <mu2> .2 </mu2>
            <material> Gazebo/Grey </material>
            <selfCollide>True</selfCollide>
        </gazebo>

        <joint name="${name}_joint" type="revolute">
            <parent link="${parent}"/>
            <child link="${name}"/>
            <origin rpy="0 0 0" xyz="${x} ${y} ${z}"/>
            <axis xyz="0 -1 0"/>
            <limit effort="100" velocity="1" lower="0" upper="3.65"/>
            <dynamics damping="1"/>
        </joint>
        <!--gazebo reference="${name}_joint">
            <stopCfm>0.8</stopCfm>
            <stopErp>0.5</stopErp>
            <provideFeedback>true</provideFeedback>
            <implicitSpringDamper>true</implicitSpringDamper>
            <fudgeFactor>0.0</fudgeFactor>
        </gazebo-->
        <gazebo>
            <plugin name="joint_limiter_${name}" filename="libJointLimiter.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <lower>0</lower>
                <upper>3.65</upper>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_encoder_${name}" filename="libJointEncoder.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/encoder/abs/${name}</topic>
                <lower>0</lower>
                <upper>3.65</upper>
                <absolute>1</absolute>
                <resolution>10000</resolution>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_controller_${name}" filename="libControllerPID.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <Kp>30</Kp>
                <Ki>3</Ki>
                <Kd>2</Kd>
            </plugin>
        </gazebo>


        <!--transmission name="${name}_transmission">
            <type>transmission_interface/SimpleTransmission</type>
            <joint name="${name}_joint">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
            </joint>
            <actuator name="${name}_motor">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
                <mechanicalReduction>1</mechanicalReduction>
            </actuator>
        </transmission-->

    </xacro:macro>
</robot>
